### ISSUE TITLE
UserAgent._sessions

### ISSUE THREAD

**Issue by InnovateAsterisk** [onsip/SIP.js#961]
**Is your feature request related to a problem? Please describe.**
UserAgent._sessions is marked as private. When switching between multiple calls, it's useful.

**Describe the solution you'd like**
UserAgent.sessions public 

**Describe alternatives you've considered**
I used ```UserAgent.sessions = UserAgent._sessions``` to make my existing code work, but I found it to be hack.

**Additional context**
I'm not sure why you have decided to depreciate so many features from older visions. It makes development with your software difficult.

**Comment by john-e-riordan** [comment]
Perhaps providing a shallow copy in an array would work? Don't want to leak a reference to the UserAgents internal session container.

**Comment by iomariani** [comment]
I use my own wrapper using the delegates provided in the UserAgent.

```ts
const sessions = new Map<string, Session>()

userAgent.delegate.onInvite = (invitation) => {
  sessions.set(invitation.id, invitation)

  invitation.stateChange.addListener((state) => {
    if (state === SessionState.Terminated) {
      sessions.delete(invitation.id)
    }
  })
}
```

**Comment by john-e-riordan** [comment]
Yes - and you can track outgoing ones in the same fashion just before calling `inviter.invite()`. Your approach is the best one currently.

The concern with exposing the sessions UserAgent is tracking is that UserAgent arguably should not be tracking them at all. The only reason UserAgent currently tracks sessions is for housekeeping when `stop()` is called and there's a case being made in another issue that UserAgent should not be doing that housekeeping (and/or what it is doing now is broken). And so if UserAgent should not be doing the housekeeping, then it would make sense to eliminate the sessions from UserAgent altogether.

All that said, sessions obviously do need to be tracked someplace if multiple sessions are being utilized.

**Comment by InnovateAsterisk** [comment]
I was previously maintaining my own sessions, but had two problems;

If the user agent maintained its own sessions, and I also maintain sessions in another object, there could be the possibility of the two becoming out of sync. I certainly don't want to maintain both.

Also I wanted to look at adding multiple registrations into my solution (that already supports multiple sessions). It would have been easier to refer to existing UserAgent objects. 

Personally I feel that if a UserAgent is responsible for creating a session, it "belongs" to the agent, and not brought into some global space. I find that messy.

**Comment by InnovateAsterisk** [comment]
I use your solution on a daily basis, and am highly focused on building a great user experience from it. It's capable of just about everything I need so far, for example: https://github.com/InnovateAsterisk/Browser-Phone/

Looking forward to hearing how you guys solve this. I'm aware of the efforts you guys put in here! So, Thanks

**Comment by john-e-riordan** [comment]
I've pushed a new class `SessionManager` to the main branch. It's basically `SimpleUser` with support for multiple concurrent sessions. I'm proposing...

- the managed sessions are public: `public managedSessions: Array<ManagedSession>`
- the `ManagedSession` interface has a reference to the associated `Session`
- the `ManagedSession` instance is created via a configurable `ManagedSessionFactory` allowing the user to supply their own custom `ManagedSession` objects
- the `SimpleUser` class is now implemented using `SessionManger`
- the library will provide for management of multiple concurrent sessions via the `SessionManager` class going forward
- the utilization of the lower level UserAgent API directly will require the user to track session on their own and the recommended approach to doing so is the one outlined earlier in this issue and codified by example in the `SessionManager` class

The upshot for `UserAgent._sessions` is that will remain private (and hopefully eliminated in the future in line with not tracking sessions in multiple places as a matter of good design). I spent a bunch of time looking at exposing the sessions at the UserAgent level this past week and my opinion is stronger now that I spent time attempting to code it up. It seems to me that managing a collection of sessions is best done "one level up" from UserAgent given the current software architecture.

So I did that and now we have `SessionManager` which suffices as the base for a number of projects I'm aware of that are currently coding to the API directly because `SimpleUser` was not sufficient. All these project are managing their own session collections and could now use `SessionManager` instead. And if one needs to manage their own sessions, the `SessionManger` class provide a working example of the recommended approach to doing so.

Feedback welcome. My default course is to push this out shortly with an upcoming 0.21.0 release.
